### ISSUE TITLE
Can't proxy websocket request using proxy option

### ISSUE THREAD

**Issue by gymb321** [webpack/webpack-dev-server#283]
The webpack-dev-server does not explicitly handle websockets through the proxy option. I would expect an entry like this to correctly proxy websocket requests:
```
proxy : [{path: /ws, target: ws://localhost:8080/, ws: true}]
```
After inspecting the Server.js file, it seem that there would need to be explicit handling of websockets. Something like this perhaps:
```
options.proxy.forEach(function (proxyOptions) {
	app.all(proxyOptions.path, function (req, res) {
		if(typeof proxyOptions.rewrite === 'function') proxyOptions.rewrite(req, proxyOptions);
		if (proxyOptions.host) {
			req.headers.host = proxyOptions.host;
		}
		if (proxyOptions.ws) {
			proxy.ws(req, res, proxyOptions, errorCb.bind(this));
		} else {
			proxy.web(req, res, proxyOptions, errorCb.bind(this));
		}
		function errorCb(err){
			var msg = "cannot proxy to " + proxyOptions.target + " (" + err.message + ")";
			this.io.sockets.emit("proxy-error", [msg]);
			res.statusCode = 502;
			res.end();
		}
		
		if (proxyOptions.configure) {
			proxyOptions.configure(proxy);
		}
	}.bind(this));
}.bind(this));
```
or add a new listener for websocket connections: 
```
proxyServer.on('upgrade', function (req, socket, head) {
  proxy.ws(req, socket, head);
});
```

**Comment by jedwards1211** [comment]
I was just investigating this, I came to the same conclusion :)

**Comment by jedwards1211** [comment]
Wait...is http-proxy supposed to automatically take care of things if I do the following?  It doesn't seem to be working with Meteor...

```js
    proxy: [
      {
        path: '*',
        ws: true,
        target: {
          host: 'localhost',
          port: 3000,
        },
      },
    ],
```

**Comment by jedwards1211** [comment]
Wait a minute...you mean you need to make a socket connection to something besides `webpack-dev-server`, right?
I still don't know a lot about web sockets but as far as I can gather, `webpack-dev-server` is in control the socket connection so one can't just simply pass through some socket traffic to another server.

**Comment by jedwards1211** [comment]
iframe mode seems to be working for me?  But I don't understand how since it's still using a socket connection to webpack-dev-server as well?

**Comment by jedwards1211** [comment]
@gymb321 I'd recommend using iframe mode (i.e. put a `contentBase` in your dev server config and go to `localhost:port/webpack-dev-server/index.html` in your browser).  This is the only way I know of to connect to a backend server via websocket.

**Comment by mingfang** [comment]
+1

**Comment by Jarema** [comment]
:+1:  Lack of websocket proxy support makes using dev-server with "real" backend servers impossible.

**Comment by readme42** [comment]
:+1:

**Comment by andreasvoigt** [comment]
:+1:

**Comment by geddski** [comment]
+1

**Comment by thomasboyt** [comment]
Looks to me that the issue with either of the suggested code snippets is that `app.all(path, cb)` in express *doesn't fire for websocket connections*, since express only listens for traditional HTTP(S) connections.

There are various middleware modules out there that try to solve this issue, such as [express-ws](https://github.com/HenningM/express-ws), but I've had trouble getting them to work (possibly because the webpack-dev-server SockJS server is interfering?).

**Comment by thomasboyt** [comment]
Hacked together a quick solution, can turn into a PR if it seems like the right route: https://github.com/thomasboyt/webpack-dev-server/commit/f2fceae96952f4d24012c747924b55cbaba06f3c

**Comment by jedwards1211** [comment]
Hey @thomasboyt, thanks so much for that!!!  I've been dying to be able to proxy websockets!

I had to make one change to get it to work; if you don't add an error handler to `proxy`, if it fires an `'error'` event, the default error handler will throw the error, which seems to break something (I'm not sure what, the process just seems to hang, but since the error is probably getting thrown in the middle of stream communication it's probably not good).  So if you simply add this underneath the requires:

```js
proxy.on('error', function(err) {
	console.error(err.stack);
});
```

It fixes that issue.  My sockets were sometimes hanging up so it was a major issue for me.

**Comment by geddski** [comment]
web sockets support will be very nice. I currently proxy them manually with a separate server.

**Comment by apotropaix** [comment]
Thanks to @sokra and @SpaceK33z (and perhaps others?) web socket proxying (with socket.io) has been working for at least a couple of weeks now since resolution #302. This is the config which is working (for me):

I have a socket.io server running on localhost:8001, on path /napi:

var io = require('socket.io')(http, { path: '/napi' })

In webpack config:
devServer: {
        proxy: {
            '/napi/*': {
                target: 'ws://localhost:8001',
                ws: true,
            },
        },
    },

In socket.io client:
var socket = io({ path: '/napi' })

I'm launching with:
webpack-dev-server --inline --hot

**Comment by jedwards1211** [comment]
@apotropaix are you saying that websocket proxying should work without @thomasboyt's patch?

**Comment by apotropaix** [comment]
@jedwards1211 Yes. confirm. I have webpack-dev-server@1.14.0 installed.

**Comment by jedwards1211** [comment]
@apotropaix I see...I'm confused though, how is that possible when there aren't any `proxy.ws()` calls in [that version's Server.js](https://github.com/webpack/webpack-dev-server/blob/v1.14.0/lib/Server.js)?

**Comment by jedwards1211** [comment]
@apotropaix nevermind, I see how it works looking at `node-http-proxy` documentation.  Thanks for enlightening us!

**Comment by jedwards1211** [comment]
@apotropaix my project uses SockJS, and I can't get it to proxy successfully, whereas @thomasboyt's patch does work.  Maybe the problem is that SockJS makes a mix of http:// and ws:// requests, but since the config can specify only one protocol in the target, one protocol will fail to get proxied?

**Comment by apotropaix** [comment]
@jedwards1211 Ah, I see, didn't catch you were using SockJS. There were similar issues with socket.io before issue 302 resolution switched the webpack-dev-server engine from socket.io to SockJS.

**Comment by SpaceK33z** [comment]
@jedwards1211, what is the path your SockJS lives at? webpack-dev-server uses `/sockjs-node`, so if you're also using this... I think we should move to using something like `/_webpack-sockjs` to prevent collisions like this.

**Comment by jedwards1211** [comment]
@SpaceK33z no, it lives at `/sockjs`, so there's no name collision.
@apotropaix in any case, luckily Thomas Boyt's patch does work with SockJS

**Comment by thomasboyt** [comment]
Ah, this is interesting.

As far as I could tell, with my client using raw websockets, webpack-dev-server didn't seem to be even handling the proxied websocket connection. I suspect that either something about SockJS or socket.io's additions over regular websockets may allow them to work where regular websockets aren't (e.g. making HTTP connections that the express server is properly handling), or I'm just doing something entirely wrong in my proxy setup. I'll set up a test case and see what I can figure out :)

**Comment by thomasboyt** [comment]
I've created a demo repo that shows that webpack-dev-server does not successfully proxy `ws://` protocol requests without the change I made on my branch: https://github.com/thomasboyt/webpack-dev-server-283-demo

It's still possible I'm doing something wrong, of course, but if I understand the docs correctly, that demo is intended to work.

**Comment by mjrussell** [comment]
I am using raw websockets as well and can confirm that the upgrade request is not being honored in webpack-dev-server@1.14.0. However I pulled down @thomasboyt's commit and it works perfectly now

**Comment by mjrussell** [comment]
Anyway to get some movement on this? @jedwards1211 @sokra? 

What do we need to do in order to get this into webpack-dev-server?

**Comment by bbguitar77** [comment]
:+1:

**Comment by chimurai** [comment]
PR #359 might just add support for `ws` proxy.
I haven't tested it yet.

**Comment by mjrussell** [comment]
@chimurai thanks, I can take your PR for a test run at some point and let you know

**Comment by chimurai** [comment]
Just connecting the dots. Credits should go to @jamsesso for making the PR.

**Comment by mjrussell** [comment]
@chimurai thanks for pointing that out. It does indeed work with #359 which is now merged to master.

**Comment by k-makarov** [comment]
:+1:

**Comment by mjrussell** [comment]
@k-makarov proxying for websockets does now work in master, however it looks like there isn't going to be a new release until webpack 2.0. I created a branch that you can directly npm install from in the meantime here: https://github.com/mjrussell/webpack-dev-server/tree/pre-dist

**Comment by geddski** [comment]
webpack-dev-server 1.14.1 has it.

**Comment by jedwards1211** [comment]
@geddski last time I checked, I couldn't get it to work with webpack 1.14...

**Comment by mjrussell** [comment]
I couldn't get it either, you need the HPM PR included which came from #359 just after 1.14.1, that's why im still using my fork until I can upgrade to 2.0

**Comment by SpaceK33z** [comment]
This has landed in webpack-dev-server `1.15.0`.

**Comment by thomasboyt** [comment]
Glad this was implemented in a `1.x` release :)

**Comment by mjrussell** [comment]
Yeah thanks for this

**Comment by davidraine** [comment]
Really great tutorial and very helpful for folks like me who need to understand WHY something is there.

Just one question. When I start up my minimalist app, I see an xhr request from sockjs and then a websocket is opened (which remains in "pending" state). Why are these occurring? I saw these before I started on the hot loading elements.  If not needed I would like to eliminate this connection.

The only change I have made from the tutorial is my webpack-dev-server listens on a different IP address/port (is not localhost).

`    "start": "webpack-dev-server --port 8080 --host 10.0.2.4 --config ./webpack.config.js --mode development",`
